### PR TITLE
fix: [DEL-3336]: Dynamically throttle task based on CPU/Memory usage

### PR DESCRIPTION
**Pull request by gaurav-jain1**
You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/30836)

### CODE REVIEW

File: 260-delegate/src/main/java/io/harness/delegate/app/modules/DelegateModule.java [hunk 1]
```diff
@@ -789,7 +789,7 @@ public ExecutorService taskProgressExecutor() {
   @Singleton
   @Named("taskExecutor")
   public ExecutorService taskExecutor() {
-    return ThreadPool.create(10, 400, 1, TimeUnit.SECONDS,
+    return ThreadPool.create(10, Integer.MAX_VALUE, 1, TimeUnit.SECONDS,
```
Comment (bharatgoelharness) [inline 260-delegate/src/main/java/io/harness/delegate/app/modules/DelegateModule.java:792]: we should have used dynamicHandlingOfRequestEnabled here also
dynamicHandlingOfRequestEnabled ? MAx_VALUE : 400

Comment (gaurav-jain1) [reply 260-delegate/src/main/java/io/harness/delegate/app/modules/DelegateModule.java:792]: Ahh Sure, make sense.

**Comment by raghuAtWings** [comment]
___
*[260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java, line 760 at r2](https://reviewable.io/reviews/harness/harness-core/30836#-MwVFUqSCRy5wEZUX0Qy:-MwVFUqTCZAzjTetVE0c:b9uvq5w) ([raw file](https://github.com/harness/harness-core/blob/d946e6fcbaaa4de52125c6b09fc04894ec5afbc2/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L760)):*
> ```Java
> 
>     if (currentRSS >= RESOURCE_USAGE_THRESHOLD * maxRSS || currentCPU >= RESOURCE_USAGE_THRESHOLD) {
>       log.error(
> ```

this should not be an error, but warning

**Comment by raghuAtWings** [comment]
___
*[960-api-services/src/main/java/io/harness/delegate/configuration/DelegateConfiguration.java, line 58 at r2](https://reviewable.io/reviews/harness/harness-core/30836#-MwVFf_G6e_r16jDjLL9:-MwVFf_G6e_r16jDjLLA:bpry0t7) ([raw file](https://github.com/harness/harness-core/blob/d946e6fcbaaa4de52125c6b09fc04894ec5afbc2/960-api-services/src/main/java/io/harness/delegate/configuration/DelegateConfiguration.java#L58)):*
> ```Java
>   private boolean clientToolsDownloadDisabled;
>   private boolean installClientToolsInBackground;
>   private boolean dynamicHandlingOfRequestEnabled;
> ```

how would existing delegate get this flag? this should be controller by a FF in manager side and when delegate get upgraded and gets the new config, we should get this set based on the FF

**Comment by raghuAtWings** [comment]
___
*[260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java, line 752 at r2](https://reviewable.io/reviews/harness/harness-core/30836#-MwVG1XM12f6ObCvYC8o:-MwVG1XM12f6ObCvYC8p:b-lg6bi4) ([raw file](https://github.com/harness/harness-core/blob/d946e6fcbaaa4de52125c6b09fc04894ec5afbc2/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L752)):*
> ```Java
>   }
> 
>   private void maybeUpdateTaskRejectionStatus() {
> ```

how expensive is this to call every 1 second?

**Comment by raghuAtWings** [comment]
___
*[260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java, line 1511 at r2](https://reviewable.io/reviews/harness/harness-core/30836#-MwVGAsd2fH9zhnLl4nL:-MwVGAsd2fH9zhnLl4nM:b-d2b5r6) ([raw file](https://github.com/harness/harness-core/blob/d946e6fcbaaa4de52125c6b09fc04894ec5afbc2/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L1511)):*
> ```Java
>         maybeUpdateTaskRejectionStatus();
>       } catch (Exception ex) {
>         log.error("Exception while sending heartbeat", ex);
> ```

why the error message says heartbeat?

**Review by gaurav-jain1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30836)* status: 3 file(s) to review. 5. unresolved discussions. 0 of 1 approvals obtained. 

___
*[260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java, line 752 at r2](https://reviewable.io/reviews/harness/harness-core/30836#-MwVG1XM12f6ObCvYC8o:-MwVdAC2DJQvShpl3RYe:bsvi0zi) ([raw file](https://github.com/harness/harness-core/blob/d946e6fcbaaa4de52125c6b09fc04894ec5afbc2/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L752)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

how expensive is this to call every 1 second?
</blockquote></details>

Underneath it will perform system call via JVM, to get the os values. But shouldn't be too expensive.
___
*[260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java, line 760 at r2](https://reviewable.io/reviews/harness/harness-core/30836#-MwVFUqSCRy5wEZUX0Qy:-MwVeuAc8EwVckTwGjE_:bs7wumb) ([raw file](https://github.com/harness/harness-core/blob/d946e6fcbaaa4de52125c6b09fc04894ec5afbc2/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L760)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

this should not be an error, but warning
</blockquote></details>

Done.
___
*[260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java, line 1511 at r2](https://reviewable.io/reviews/harness/harness-core/30836#-MwVGAsd2fH9zhnLl4nL:-MwVfK2U9DQ8sumYeB2P:b-ebxnxx) ([raw file](https://github.com/harness/harness-core/blob/d946e6fcbaaa4de52125c6b09fc04894ec5afbc2/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L1511)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

why the error message says heartbeat?
</blockquote></details>

Oops my bad.. typo error.
___
*[960-api-services/src/main/java/io/harness/delegate/configuration/DelegateConfiguration.java, line 58 at r2](https://reviewable.io/reviews/harness/harness-core/30836#-MwVFf_G6e_r16jDjLL9:-MwVfhfg3zWCr0T8MmLQ:b-19bc9c) ([raw file](https://github.com/harness/harness-core/blob/d946e6fcbaaa4de52125c6b09fc04894ec5afbc2/960-api-services/src/main/java/io/harness/delegate/configuration/DelegateConfiguration.java#L58)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

how would existing delegate get this flag? this should be controller by a FF in manager side and when delegate get upgraded and gets the new config, we should get this set based on the FF
</blockquote></details>

I haven't think how to enable this as off now. Introduced this just to make sure this code is currently disabled and we can internally use this for internal testing. Once we are sure, we can probably rolled this out based on FF or, remove this flag completely.

**Comment by gaurav-jain1** [comment]
trigger ti0

**Review by raghuAtWings — COMMENTED** [review]
> Reviewed 1 of 10 files at r4, 6 of 7 files at r5, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30836)* status: 2. unresolved discussions. 0 of 1 approvals obtained.

**Review by raghuAtWings — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30836)* status: 1. unresolved discussions.

**Comment by gaurav-jain1** [comment]
trigger ti1

**Pull request merged**